spring/boot3-autoconfigure/src/main/java/com/linecorp/armeria/spring/InternalServices.java

```diff
@@ -60,6 +62,18 @@ private static boolean hasAllClasses(String... classNames) {
         return true;
     }
 
+    static {
+        // InternalServices is the only class that both boot-starter and boot-webflux-starter always depend on.
+
+        // Disable the default shutdown hook to gracefully close the client factory after the server is
+        // shut down.
+        ClientFactory.disableShutdownHook();
+        // The shutdown hooks are invoked after all other contexts are closed.
+        // The server is closed by ConfigurableApplicationContext.closeAndWait().
+        // https://github.com/spring-projects/spring-boot/blame/781d7b0394c71e20f098f64a3261a18346ccd915/spring-boot-project/spring-boot/src/main/java/org/springframework/boot/SpringApplicationShutdownHook.java#L114-L116
+        SpringApplication.getShutdownHandlers().add(ClientFactory::closeDefault);
```

Comment:
Actually, I realized that `Spring` uses an `IdentityHashMap` internally which doesn't really guarantee insertion-order iteration

Reply:
That wouldn't be a problem for our cases because happen-before is guaranteed for them. A shutdown hook for a server is stored separately at `contexts` map which is executed before the hooks added with `getShutdownHandlers.add()`.

Reply:
Double checked that the hook is executed after all contexts are closed.
```
@Override
public void run() {
	Set<ConfigurableApplicationContext> contexts;
	Set<ConfigurableApplicationContext> closedContexts;
	Set<Runnable> actions;
	synchronized (SpringApplicationShutdownHook.class) {
		this.inProgress = true;
		contexts = new LinkedHashSet<>(this.contexts);
		closedContexts = new LinkedHashSet<>(this.closedContexts);
		actions = new LinkedHashSet<>(this.handlers.getActions());
	}
	contexts.forEach(this::closeAndWait);
	closedContexts.forEach(this::closeAndWait);
	actions.forEach(Runnable::run);
}
```

Reply:
I missed that point 👍  Thanks for the explanation